lib/linked_list.rb

```diff
@@ -19,99 +19,229 @@ def initialize
+        
+      until current.nil?
+        current = current.next
+        length += 1
+      end
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      current = @head
+      current_index = 0
+        
+      index.times do |i|
+        return nil if current.nil?
+        current = current.next
+      end
+      
+      return current.data
     end
 
     # method to print all the values in the linked list
     def visit
-      raise NotImplementedError
+      current = @head
+
+      until current.nil?
+        print current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
     def delete(value)
-      raise NotImplementedError
+      current = @head
+      previous = nil
+
+      return nil if current.nil?
+      until current.data == value
+        previous = current
+        current = current.next
+      end
+
+      if previous.nil?
+        @head = current.next
+      else
+        previous.next = current.next
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
     def reverse
-      raise NotImplementedError
+      current = @head
+      previous = nil
+
+      until current.nil?
+        temp = current
+        current = current.next
+        temp.next = previous
+        previous = temp
+      end
+      @head = previous
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
     def find_middle_value
-      raise NotImplementedError
+      current = @head
+      middle = current
+
+      return nil if current.nil?
+      until current.nil?
+        middle.next = middle
+        current.next.next = current
+      end
+      return middle
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     def find_nth_from_end(n)
-      raise NotImplementedError
+      current = @head
+      from_end = current
+
+      return nil if current.nil?
+
+      n.times do
+        return nil if current.next.nil?
+        current = current.next
+      end
+      
+      until current.next.nil?
+        return from_end.data if from_end.next.nil?
+        from_end = from_end.next
+        current = current.next
+      end
+      return from_end.data
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     def has_cycle
-      raise NotImplementedError
+      slow = @head
+      fast = @head
+
+      until fast.nil?
+        slow = slow.next
+        fast = fast.next.next

```

Comment:
Just in case fast.next is nil.

```suggestion
        fast = fast.next.nil? ? fast.next : fast.next.next
```